[api-server] Object store folder path is configurable and can work with AWS (secure and region flag, and IAM credentials)

eterna2 opened the pull request:
### Background

Current, ml-pipeline api server hardcodes the folder to store the pipeline templates. This disrupts how I can deploy multiple kubeflow deployments especially for productions.

I controls access via the env prefix (dev, stg, prd), and the hardcoded pipelineFolder forces me to control access via diff buckets instead of paths, and also prevents me from putting the pipelines templates together with the artifacts (artifacts path are customizable).

### Changes (Updated 16 Jan 2020)

- made `ObjectStore` folder path configurable from both config.json and env
- updated `minioclient` to use chained provider credentials: API key -> minio env flags -> aws env flags -> aws IAM
- `minioclient` `secure` and `region` flags can now also be set via `config.json` and env var.

```
	minioServiceHost       = "MINIO_SERVICE_SERVICE_HOST"
	minioServicePort       = "MINIO_SERVICE_SERVICE_PORT"
	minioServiceRegion     = "MINIO_SERVICE_REGION"
	minioServiceSecure     = "MINIO_SERVICE_SECURE"
	pipelineBucketName     = "MINIO_PIPELINE_BUCKET_NAME"
	pipelineFolder         = "MINIO_PIPELINE_FOLDER"
``` 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/2080)
<!-- Reviewable:end -->

Ark-kun commented:
/ok-to-test

eterna2 commented:
ok. rebased to latest master, and fixed most of the bugs. Added a simple unit test for the new objectPath type.

IronPan commented on backend/src/apiserver/client_manager.go:
```diff
@@ -63,6 +67,7 @@ type ClientManager struct {
 	dBStatusStore          storage.DBStatusStoreInterface
 	defaultExperimentStore storage.DefaultExperimentStoreInterface
 	objectStore            storage.ObjectStoreInterface
+	objectPaths        storage.ObjectPathsInterface
```
please go fmt the code.

IronPan replied:
creating an interface for customizing the path seems a bit overkilling. 
can we add a root path in MinioObjectStore and set the path when initializing it? When add/get file we just append root path to the filepath?

eterna2 commented:
hey @IronPan 
sorry for the long delay. Fixed the PR as requested.

- Added `PipelineFolder` to the config.
- Added `MINIO_PIPELINE_BUCKET_NAME` and `MINIO_PIPELINE_FOLDER` env var for consistency
- CreatePipelinePath is removed as a static func, and is now a method insider MinioObjectStore
- AddFile, DeleteFile, etc all will automatically append the base folder (PipelineFolder)
- Updated unit test to reflect these changes

eterna2 commented:
Updates to objectstore and corresponding unit test:

### Pipeline folder
- Added `ObjectStoreConfig.PipelineFolder` to the config.
- Added `MINIO_PIPELINE_BUCKET_NAME` and `MINIO_PIPELINE_FOLDER` env var for consistency
- `CreatePipelinePath` is removed as a static func, and is now a method insider `MinioObjectStore`
- `AddFile`, `DeleteFile`, etc all will automatically append the base folder (`PipelineFolder`)

### Fix #2819 : Region for AWS access (i.e. S3)
- Added `ObjectStoreConfig.Region` to the config.
- added `MINIO_SERVICE_REGION` env flag to set region to use for minio client.

### Duplicated with #2720: Secure connection with AWS S3 endpoint
- Added `ObjectStoreConfig.Secure` to the config.
- added `MINIO_SERVICE_SECURE` env flag to set region to use for minio client.

PS: @descampsk, hope u don't mind. Because all these 3 feature requests are so closely related, and I just happens to see ur PR and the CI errors. U need to install bazel in ur local machine to do the build and run the test. U can't see the detailed error logs from the CI - u need to run bazel locally and see the logs. I am not sure what is ur error thou - ur PR looks fine to me.

descampsk commented:
@eterna2 
Hello,
No worries for me, I'm glad you added that feature I was trying to push.
If it also allows to merge it faster, that's just a bonus.
I'm leaving my PR open for now, and I'll close it once yours is merged.

descampsk commented:
Will it work with minioHost = s3.amazonaws.com, minioPort=443 et minioSecure=true ?

eterna2 commented:
Yeah. It should work. Cuz under the hood for minio-go it just parses the final endpoint to `https://s3.awsamazon.com:443`

eterna2 commented:
### Changes (Updated 16 Jan 2020)

- made `ObjectStore` folder path configurable from both config.json and env
- updated `minioclient` to use chained provider credentials: API key -> minio env flags -> aws env flags -> aws IAM
- `minioclient` `secure` and `region` flags can now also be set via `config.json` and env var.

```
	minioServiceHost       = "MINIO_SERVICE_SERVICE_HOST"
	minioServicePort       = "MINIO_SERVICE_SERVICE_PORT"
	minioServiceRegion     = "MINIO_SERVICE_REGION"
	minioServiceSecure     = "MINIO_SERVICE_SECURE"
	pipelineBucketName     = "MINIO_PIPELINE_BUCKET_NAME"
	pipelineFolder         = "MINIO_PIPELINE_FOLDER"
```

eterna2 commented:
This PR should also fixes https://github.com/e2fyi/kubeflow-aws/issues/1

eterna2 commented:
TODO need to check `ResourceManager.ReadArtifact` calls to `r.objectStore.GetFile(artifactPath)` as the PR introduces a default base folder to the objectStore API.

Jeffwan commented on backend/src/apiserver/client/minio.go:
```diff
@@ -16,31 +16,61 @@ package client
+}
+
+// createCredentialProvidersChain creates a chained providers credential for a minio client
+func createCredentialProvidersChain(endpoint, accessKey, secretKey string) *credentials.Credentials {
```
seems endpoint is not being used. could we clean it up?

discordianfish replied:
Any reason this wasn't addressed? Given it's a new contribution I wouldn't expect to find such cruft.

Jeffwan commented on backend/src/apiserver/client_manager.go:
```diff
@@ -37,6 +37,10 @@ import (
 	minioServicePort       = "MINIO_SERVICE_SERVICE_PORT"
+	minioServiceRegion     = "MINIO_SERVICE_REGION"
+	minioServiceSecure     = "MINIO_SERVICE_SECURE"
+	pipelineBucketName     = "MINIO_PIPELINE_BUCKET_NAME"
```
Any reason you add this environment?  Does config override work for you?  If so, it's unnecessary to add a new env for same field

eterna2 replied:
true. this is more for consistency.  Personally, I feel we shld remove all the `MINIO_*` and stick with what `viper` provides us.

eterna2 replied:
The only reason i prefer `env` over `config` is just it takes less effort than attaching a configmap.

Jeffwan commented on backend/src/apiserver/client_manager.go:
```diff
@@ -309,29 +313,34 @@ func initMinioClient(initConnectionTimeout time.Duration) storage.ObjectStoreInt
+	accessKey := common.GetStringConfigWithDefault("ObjectStoreConfig.AccessKey", "")
+	secretKey := common.GetStringConfigWithDefault("ObjectStoreConfig.SecretAccessKey", "")
+	bucketName := common.GetStringConfigWithDefault("ObjectStoreConfig.BucketName", os.Getenv(pipelineBucketName))
+	pipelineFolder := common.GetStringConfigWithDefault("ObjectStoreConfig.PipelineFolder", os.Getenv(pipelineFolder))
```
I am thinking maybe we can change `pipelineFolder`  to `pipelinePath` ? object key could be multi-folder like `folderA/folderB/object_name`

eterna2 replied:
ok

Jeffwan commented:
@eterna2 Trying to get better understand of your case. 

> Current, ml-pipeline api server hardcodes the folder to store the pipeline templates. This disrupts how I can deploy multiple kubeflow deployments especially for productions.

Do you mean you have multiple kubeflow clusters for one env and you want to share same bucket and path for pipeline data? 

> I controls access via the env prefix (dev, stg, prd), and the hardcoded pipelineFolder forces me to control access via diff buckets instead of paths

I think it's expected and suggested, different bucket probably want to have different policy. Do you mean you want to use same bucket and have different folder for different env? like
```
bucket/dev-pipeline
bucket/stg-pipeline
bucket/prd-pipeline
```

> and also prevents me from putting the pipelines templates together with the artifacts (artifacts path are customizable).

Any reason you want to put pipeline with artifacts folder together?

eterna2 commented:
> @eterna2 Trying to get better understand of your case.
> 
> > Current, ml-pipeline api server hardcodes the folder to store the pipeline templates. This disrupts how I can deploy multiple kubeflow deployments especially for productions.
> 
> Do you mean you have multiple kubeflow clusters for one env and you want to share same bucket and path for pipeline data?

I am running multiple kubeflow pipelines in 2 clusters. In different namespace for different teams for dev and stg/prd - i.e. each team will have 2 kubeflow pipelines, 1 dev and 1 stg/prd. And we have several teams.

So currently the `prefix` to save the pipeline templates are hardcoded. I have 1 s3 bucket for each cluster. And within a cluster, different namespace saves to a different prefix (by team). So if the pipeline path is not configurable, I will need to give each team 2 s3 buckets (dev and stg/prd). 

tldr. same bucket for pipelines in same cluster but different path for each pipeline in different namespace.

> 
> > I controls access via the env prefix (dev, stg, prd), and the hardcoded pipelineFolder forces me to control access via diff buckets instead of paths
> 
> I think it's expected and suggested, different bucket probably want to have different policy. Do you mean you want to use same bucket and have different folder for different env? like
> 
> ```
> bucket/dev-pipeline
> bucket/stg-pipeline
> bucket/prd-pipeline
> ```

yup. But not dev/stg/prd, rather by teamA/teamB/teamC. I use diff buckets for dev/stg/prd.

> 
> > and also prevents me from putting the pipelines templates together with the artifacts (artifacts path are customizable).
> 
> Any reason you want to put pipeline with artifacts folder together?

cuz they are tightly coupled. 

I put everything related to a specific kf pipeline deployment in the same folder. i.e. the s3 looks something like:

```
- dev-bucket
  - pipelines
    - namespace A
      - artifacts/**/*/main.log
      - pipelines/**
    - namespace B
      - artifacts/**/*/main.log
      - pipelines/**

```

eterna2 commented:
ok. I have rebase and sync with the current master. Removed `getEndpoint` cuz I saw the new master has `joinHostPort` which is the same thing.

Renamed `pipelineFolder` to `pipelinePath`. Updated all the unit tests again.

I have tested this on my own cluster and it works fine. 

Pending the decision on the extra env flags. Otherwise shld be good to go.

IronPan commented:
/lgtm

IronPan commented:
/approve

Jeffwan commented:
Thanks!

descampsk commented:
Thanks it will be very helpful !

discordianfish commented on backend/src/apiserver/client_manager.go:
```diff
@@ -317,29 +321,34 @@ func initMinioClient(initConnectionTimeout time.Duration) storage.ObjectStoreInt
-func createMinioBucket(minioClient *minio.Client, bucketName string) {
+func createMinioBucket(minioClient *minio.Client, bucketName, region string) {
+	// Check to see if we already own this bucket.
+	exists, err := minioClient.BucketExists(bucketName)
```
Error isn't checked

eterna2 commented:
@discordianfish

The initial PR was opened many months ago before the `joinHostPort` func was merged into master.

There was a large desync from master when the review was done, hence it seems to be "unused". But it was resolved - i.e. removed, after I rebased with master - as commented after the rebase.

U r probably right for the No error checked for bucket exist cuz I am lazy. I am catching the error only after I try creating the bucket.

Mostly because exist can only be true if there is no error. If it does not exist, or the check fails for any reason, I am going to just try creating the bucket to see if it works. If it fails, then I raise an error.
